## In this notebook, I am going to write a python script to merge the "MetroHospitalList_Metro.csv" and "" file to get the Telephone number of all the Hospitals in Victoria because it will be used in the Appliaction.

## Then I will again merge the new dataset with "Hospital_Locations.csv" to get the X and Y co-ordinates of the Hospitals. 

## The main goal is to Merge the 3 Big datasets into one Large datasets which will have Longitude, Latitude, Address, Postcode and Telephone number of each Hospital in Victoria. So that, it will easier to use the dataset in the app to fetch required infomation about Hospitals.

First, let's import the necessary libraries

In [1]:
import pandas as pd                                 # To manipulate dataframes
import re                                           # For using Regular Expressions
import warnings                                     # To ignore the warnings in order to beautify the script
warnings.filterwarnings('ignore')

Now, Let's import the **"MetroHospitalList_Metro.csv"** datatset and store it as **metro** dataframe

In [2]:
metro = pd.read_csv("MetroHospitalList_Metro.csv")            

Let's look at first few rows of the dataset to have an idea

In [3]:
metro.head()

ParentName         Hospital Name  \
0  Alfred Health         Alfred Health   
1  Alfred Health    Caulfield Hospital   
2  Alfred Health  Sandringham Hospital   
3  Alfred Health            The Alfred   
4  Austin Health         Austin Health   

                                             Address     Location  Postcode  \
0  c/- The Alfred, 55 Commercial Road Melbourne V...    Melbourne      3004   
1                260 Kooyong Road Caulfield VIC 3162    Caulfield      3162   
2            193 Bluff Road\r\r Sandringham VIC 3191  Sandringham      3191   
3          55 Commercial Road \r\rMelbourne VIC 3004    Melbourne      3004   
4           145 Studley Road \r\rHeidelberg VIC 3084   Heidelberg      3084   

    CEO/EO Site               Hospital Email     Telephone           Fax  \
0    Andrew Way            ceo@alfred.org.au  03 9076 2000  03 9076 3409   
1    Andrew Way  publicaffairs@alfred.org.au  03 9076 6000  03 9076 6434   
2    Andrew Way  publicaffairs@alfred.org.au  03 9076 1000  03 9598 1539   
3    Andrew Way  publicaffairs@alfred.org.au  03 9076 2000  03 9076 2222   
4  Sue Shilbury    marie.ellis@austin.org.au  03 9496 5000  03 9458 4779   

  RegistrarTypeCode  
0                AU  
1                AG  
2                AG  
3                AG  
4                AU

It's time to import the second dataset **"RuralHospitalList_rural.csv"** and store it as **rural** dataframe

In [4]:
rural = pd.read_csv("RuralHospitalList_rural.csv")

Let's have a look into the head of the rural dataset.

In [5]:
rural.head()

Hospital Name  \
0               Albury Wodonga Health   
1           Alexandra District Health   
2                       Alpine Health   
3  Bairnsdale Regional Health Service   
4            Ballarat Health Services   

                                       Address    Location  Postcode  \
0       69 Vermont Street\r\r Wodonga VIC 3690     Wodonga      3690   
1          20 Cooper Street Alexandra VIC 3714   Alexandra      3714   
2  30 O'Donnell Avenue \r\rMyrtleford VIC 3737  Myrtleford      3737   
3           122 Day St \r\rBairnsdale VIC 3875  Bairnsdale      3875   
4  Drummond Street North \r\rBallarat VIC 3353    Ballarat      3353   

     CEO/EO Site                   Hospital Email     Telephone           Fax  
0  Janet Chapman                   ceo@awh.org.au  02 6058 4780  02 6058 4680  
1  Debbie Rogers      alexandra@humehealth.org.au  03 5772 0900  03 5772 0919  
2    Lyndon Seys  lyndon.seys@alpinehealth.org.au  03 5751 9300  03 5751 9396  
3   Robyn Hayles                email@brhs.com.au  03 5150 3333  03 5150 3367  
4    Dale Fraser                  info@bhs.org.au  03 5320 4000  03 5320 4828

#### Now, let's merge the two datasets into one

In [6]:
df = pd.concat([metro, rural], ignore_index=True)

#### Let's have a look at the newly created datafram

In [7]:
df.head()

Address   CEO/EO Site  \
0  c/- The Alfred, 55 Commercial Road Melbourne V...    Andrew Way   
1                260 Kooyong Road Caulfield VIC 3162    Andrew Way   
2            193 Bluff Road\r\r Sandringham VIC 3191    Andrew Way   
3          55 Commercial Road \r\rMelbourne VIC 3004    Andrew Way   
4           145 Studley Road \r\rHeidelberg VIC 3084  Sue Shilbury   

            Fax               Hospital Email         Hospital Name  \
0  03 9076 3409            ceo@alfred.org.au         Alfred Health   
1  03 9076 6434  publicaffairs@alfred.org.au    Caulfield Hospital   
2  03 9598 1539  publicaffairs@alfred.org.au  Sandringham Hospital   
3  03 9076 2222  publicaffairs@alfred.org.au            The Alfred   
4  03 9458 4779    marie.ellis@austin.org.au         Austin Health   

      Location     ParentName  Postcode RegistrarTypeCode     Telephone  
0    Melbourne  Alfred Health      3004                AU  03 9076 2000  
1    Caulfield  Alfred Health      3162                AG  03 9076 6000  
2  Sandringham  Alfred Health      3191                AG  03 9076 1000  
3    Melbourne  Alfred Health      3004                AG  03 9076 2000  
4   Heidelberg  Austin Health      3084                AU  03 9496 5000

#### There are some unnecessary columns. Now, let's remove those columns for simplification.

In [8]:
df= df.drop(['ParentName', 'RegistrarTypeCode','CEO/EO Site','Address','Hospital Email','Fax','Location'], axis=1)

### We have removed the above columns because our main goal is to get the Hospital Name, Postcode and Telephone number from these datasets and Location and Address will be generated from the next big dataset

#### Let's have a look in the new dataset

In [9]:
df.head()

Hospital Name  Postcode     Telephone
0         Alfred Health      3004  03 9076 2000
1    Caulfield Hospital      3162  03 9076 6000
2  Sandringham Hospital      3191  03 9076 1000
3            The Alfred      3004  03 9076 2000
4         Austin Health      3084  03 9496 5000

Let's rename the column **'Hospital Name'** into **'HospitalName'** for simplification in the later part

In [10]:
df = df.rename(columns={'Hospital Name': 'HospitalName'})

In [11]:
df.head()

HospitalName  Postcode     Telephone
0         Alfred Health      3004  03 9076 2000
1    Caulfield Hospital      3162  03 9076 6000
2  Sandringham Hospital      3191  03 9076 1000
3            The Alfred      3004  03 9076 2000
4         Austin Health      3084  03 9496 5000

#### Let's sort the dataframe according to Hospital names

In [12]:
df.sort_values(by=['HospitalName'], inplace=True)

In [13]:
df.head()

HospitalName  Postcode     Telephone
56      Albury Wodonga Health      3690  02 6058 4780
57  Alexandra District Health      3714  03 5772 0900
0               Alfred Health      3004  03 9076 2000
58              Alpine Health      3737  03 5751 9300
11           Angliss Hospital      3156  03 9764 6111

#### Now, let's reset the index 

In [14]:
df = df.reset_index(drop=True)

In [15]:
df.head()

HospitalName  Postcode     Telephone
0      Albury Wodonga Health      3690  02 6058 4780
1  Alexandra District Health      3714  03 5772 0900
2              Alfred Health      3004  03 9076 2000
3              Alpine Health      3737  03 5751 9300
4           Angliss Hospital      3156  03 9764 6111

#### Next, check if there is any null values in the dataframe

In [16]:
df.isnull().sum()

HospitalName    0
Postcode        0
Telephone       0
dtype: int64

#### There is no null value which is really good news. Now, let's check the type of the columns

In [17]:
df.dtypes

HospitalName    object
Postcode         int64
Telephone       object
dtype: object

### Up next, I will import the "Hospital_Locations.csv" and store it as "df_new" dataframe and do some cleaning task on it

In [18]:
df_new = pd.read_csv("Hospital_Locations.csv")

#### Let's have a look at the dataset

In [19]:
df_new.head()

X          Y  FID                                    LabelName  \
0  144.983399 -37.809681    1       East Melbourne Specialist Day Hospital   
1  145.149300 -37.951400    2                   Windsor Avenue Day Surgery   
2  145.344653 -38.034580    3       Hyperbaric Health Wound Centre Berwick   
3  143.832286 -37.540553    4                Ballarat Day Procedure Centre   
4  145.143379 -38.148189    5  Bayside Day Procedure and Specialist Centre   

                         OpsName     Type  StreetNum    RoadName RoadType  \
0    EAST MELBOURNE DAY HOSPITAL  PRIVATE         23   CLARENDON   STREET   
1     WINDSOR AVENUE DAY SURGERY  PRIVATE         17     WINDSOR   AVENUE   
2    HYPERBARIC HEALTH - BERWICK  PRIVATE          3        GIBB   STREET   
3  BALLARAT DAY PROCEDURE CENTRE  PRIVATE  1117-1123      HOWITT   STREET   
4   BAYSIDE DAY PROCEDURE CENTRE  PRIVATE        141  CRANBOURNE     ROAD   

  RoadSuffix  CampusCode            LGAName LocalityNa  Postcode VicgovRegi  \
0                   8280          MELBOURNE                 3002       NWMR   
1                   7050  GREATER DANDENONG                 3171        SMR   
2                   7100              CASEY                 3806        SMR   
3                   8430           BALLARAT                 3355       Gram   
4                   7310          FRANKSTON                 3199        SMR   

  State ServiceNam  
0   VIC             
1   VIC             
2   VIC             
3   VIC             
4   VIC

#### It is large dataset so let's see it's dimensions 

In [20]:
df_new.shape

(325, 17)

#### It has 325 rows and 17 columns. So, let's remove the unncessary columns first

In [21]:
df_new = df_new.drop(['FID', 'RoadSuffix','CampusCode','VicgovRegi','State','Type','ServiceNam','OpsName'], axis=1)

#### After removing the columns let's have a look at the new dataframe

In [22]:
df_new.head()

X          Y                                    LabelName  \
0  144.983399 -37.809681       East Melbourne Specialist Day Hospital   
1  145.149300 -37.951400                   Windsor Avenue Day Surgery   
2  145.344653 -38.034580       Hyperbaric Health Wound Centre Berwick   
3  143.832286 -37.540553                Ballarat Day Procedure Centre   
4  145.143379 -38.148189  Bayside Day Procedure and Specialist Centre   

   StreetNum    RoadName RoadType            LGAName LocalityNa  Postcode  
0         23   CLARENDON   STREET          MELBOURNE                 3002  
1         17     WINDSOR   AVENUE  GREATER DANDENONG                 3171  
2          3        GIBB   STREET              CASEY                 3806  
3  1117-1123      HOWITT   STREET           BALLARAT                 3355  
4        141  CRANBOURNE     ROAD          FRANKSTON                 3199

#### The next step is to rename the column 'LabelName' to 'HospitalName' to match the previous dataset

In [23]:
df_new = df_new.rename(columns={'LabelName': 'HospitalName'})

#### Up next, let's sort the dataframe according to the Hospital names

In [24]:
df_new.sort_values(by=['HospitalName'], inplace=True)

In [25]:
df_new.head()

X          Y                     HospitalName StreetNum  \
93   144.972137 -37.834298               Albert Road Clinic     31-33   
319  146.939700 -36.078500   Albury Wodonga Health (Albury)       201   
311  146.879500 -36.131800  Albury Wodonga Health (Wodonga)     53-81   
202  145.716100 -37.195300      Alexandra District Hospital        12   
203  146.966100 -36.733000           Alpine Health (Bright)     36-42   

    RoadName RoadType          LGAName         LocalityNa  Postcode  
93    ALBERT     ROAD     PORT PHILLIP                         3004  
319  BORELLA     ROAD       Albury (C)  EAST ALBURY (NSW)      2640  
311  VERMONT   STREET     Wodonga (RC)            WODONGA      3690  
202   COOPER   STREET  Murrindindi (S)          ALEXANDRA      3714  
203   COBDEN   STREET       Alpine (S)             BRIGHT      3741

#### Reset the index of the dataframe

In [26]:
df_new = df_new.reset_index(drop=True)

In [27]:
df_new.head()

X          Y                     HospitalName StreetNum RoadName  \
0  144.972137 -37.834298               Albert Road Clinic     31-33   ALBERT   
1  146.939700 -36.078500   Albury Wodonga Health (Albury)       201  BORELLA   
2  146.879500 -36.131800  Albury Wodonga Health (Wodonga)     53-81  VERMONT   
3  145.716100 -37.195300      Alexandra District Hospital        12   COOPER   
4  146.966100 -36.733000           Alpine Health (Bright)     36-42   COBDEN   

  RoadType          LGAName         LocalityNa  Postcode  
0     ROAD     PORT PHILLIP                         3004  
1     ROAD       Albury (C)  EAST ALBURY (NSW)      2640  
2   STREET     Wodonga (RC)            WODONGA      3690  
3   STREET  Murrindindi (S)          ALEXANDRA      3714  
4   STREET       Alpine (S)             BRIGHT      3741

#### In the next step, I will run some code to remove the extra names in the Brackets from the HospitalName column by using regular expressions.

In [28]:
for index,value in enumerate(df_new.HospitalName):
    new_value = re.sub(r"\(.*\)","",value)
    df_new.HospitalName[index] = new_value

In [29]:
df_new.head()

X          Y                 HospitalName StreetNum RoadName  \
0  144.972137 -37.834298           Albert Road Clinic     31-33   ALBERT   
1  146.939700 -36.078500       Albury Wodonga Health        201  BORELLA   
2  146.879500 -36.131800       Albury Wodonga Health      53-81  VERMONT   
3  145.716100 -37.195300  Alexandra District Hospital        12   COOPER   
4  146.966100 -36.733000               Alpine Health      36-42   COBDEN   

  RoadType          LGAName         LocalityNa  Postcode  
0     ROAD     PORT PHILLIP                         3004  
1     ROAD       Albury (C)  EAST ALBURY (NSW)      2640  
2   STREET     Wodonga (RC)            WODONGA      3690  
3   STREET  Murrindindi (S)          ALEXANDRA      3714  
4   STREET       Alpine (S)             BRIGHT      3741

#### There are some empty spaces in the names of the Hospital. Let's remove the empty spaces from the beginning and end part

In [30]:
for index,value in enumerate(df_new.HospitalName):
    new_value = value.strip()
    df_new.HospitalName[index] = new_value

#### Similarly, do the same in the earlier dataset

In [31]:
for index,value in enumerate(df.HospitalName):
    new_value = value.strip()
    df.HospitalName[index] = new_value

#### Now, let's merge the two datasets on the columns 'HospitalName' and 'Postcode' to get one large dataset named hospital

In [32]:
hospital = pd.merge(df_new,df, on=['HospitalName','Postcode'], how = 'left')

#### Left outer join has been implemented. Now, Let's look at the tail of the newly created dataframe

In [33]:
hospital.tail()

X          Y                        HospitalName StreetNum  \
320  144.700757 -37.887037                      Wyndham Clinic      242A   
321  145.351400 -37.756600                 Yarra Ranges Health        25   
322  146.678100 -38.557100  Yarram and District Health Service     85-91   
323  146.006833 -36.011198                   Yarrawonga Health        33   
324  145.431300 -37.213900  Yea and District Memorial Hospital        45   

       RoadName RoadType           LGAName  LocalityNa  Postcode     Telephone  
320     HOPPERS     LANE           WYNDHAM                  3030           NaN  
321      MARKET   STREET  Yarra Ranges (S)    LILYDALE      3140  03 9091 8888  
322  COMMERCIAL     ROAD    Wellington (S)      YARRAM      3971  03 5182 0222  
323       PIPER   STREET         Moira (S)  YARRAWONGA      3730  03 5743 8111  
324     STATION   STREET   Murrindindi (S)         YEA      3717  03 5736 0400

### The columns 'LGAName' and 'LocalityNa' mainly means the suburb names of the Hospitals. The 'LocalityNa'  column is more accurate than the LGAName. Therefore I will replace the value of LGAName column with non null LocalityNa values

In [34]:
locality_index = []
for index,value in enumerate(hospital.LocalityNa):
    if value != " ":
        locality_index.append(index)

In [35]:
for each in locality_index:
    hospital.LGAName[each] = hospital.LocalityNa[each]

#### Let's have a look into the new dataframe

In [36]:
hospital.tail()

X          Y                        HospitalName StreetNum  \
320  144.700757 -37.887037                      Wyndham Clinic      242A   
321  145.351400 -37.756600                 Yarra Ranges Health        25   
322  146.678100 -38.557100  Yarram and District Health Service     85-91   
323  146.006833 -36.011198                   Yarrawonga Health        33   
324  145.431300 -37.213900  Yea and District Memorial Hospital        45   

       RoadName RoadType     LGAName  LocalityNa  Postcode     Telephone  
320     HOPPERS     LANE     WYNDHAM                  3030           NaN  
321      MARKET   STREET    LILYDALE    LILYDALE      3140  03 9091 8888  
322  COMMERCIAL     ROAD      YARRAM      YARRAM      3971  03 5182 0222  
323       PIPER   STREET  YARRAWONGA  YARRAWONGA      3730  03 5743 8111  
324     STATION   STREET         YEA         YEA      3717  03 5736 0400

### Lastly, I will make a new column named Address which will have the address of each hospital in a correct format.

####  The correct format should be - StreetNum+RoadName+RoadType+LGAName

In [37]:
address = []
for each in range(len(hospital)):
    add = hospital.StreetNum[each]+" "+hospital.RoadName[each]+" "+hospital.RoadType[each]+","+hospital.LGAName[each]
    address.append(add)

#### Add the created list as Address column in the hospital dataframe

In [38]:
hospital['Address']=address

#### Let's have a look at the dataframe

In [39]:
hospital.tail()

X          Y                        HospitalName StreetNum  \
320  144.700757 -37.887037                      Wyndham Clinic      242A   
321  145.351400 -37.756600                 Yarra Ranges Health        25   
322  146.678100 -38.557100  Yarram and District Health Service     85-91   
323  146.006833 -36.011198                   Yarrawonga Health        33   
324  145.431300 -37.213900  Yea and District Memorial Hospital        45   

       RoadName RoadType     LGAName  LocalityNa  Postcode     Telephone  \
320     HOPPERS     LANE     WYNDHAM                  3030           NaN   
321      MARKET   STREET    LILYDALE    LILYDALE      3140  03 9091 8888   
322  COMMERCIAL     ROAD      YARRAM      YARRAM      3971  03 5182 0222   
323       PIPER   STREET  YARRAWONGA  YARRAWONGA      3730  03 5743 8111   
324     STATION   STREET         YEA         YEA      3717  03 5736 0400   

                          Address  
320     242A HOPPERS LANE,WYNDHAM  
321     25 MARKET STREET,LILYDALE  
322  85-91 COMMERCIAL ROAD,YARRAM  
323    33 PIPER STREET,YARRAWONGA  
324         45 STATION STREET,YEA

#### Lastly, I will remove the redundant columns from the dataframe

In [40]:
hospital = hospital.drop(['StreetNum','RoadName','RoadType','LocalityNa','LGAName'], axis=1)

#### Let's have a look at the final dataframe

In [41]:
hospital.tail()

X          Y                        HospitalName  Postcode  \
320  144.700757 -37.887037                      Wyndham Clinic      3030   
321  145.351400 -37.756600                 Yarra Ranges Health      3140   
322  146.678100 -38.557100  Yarram and District Health Service      3971   
323  146.006833 -36.011198                   Yarrawonga Health      3730   
324  145.431300 -37.213900  Yea and District Memorial Hospital      3717   

        Telephone                       Address  
320           NaN     242A HOPPERS LANE,WYNDHAM  
321  03 9091 8888     25 MARKET STREET,LILYDALE  
322  03 5182 0222  85-91 COMMERCIAL ROAD,YARRAM  
323  03 5743 8111    33 PIPER STREET,YARRAWONGA  
324  03 5736 0400         45 STATION STREET,YEA

In [42]:
hospital.shape

(325, 6)

### The dataset has 325 rows and 6 columns . Now let's export the dataset as a csv file.

In [43]:
hospital.to_csv("hospitals.csv", index = False, header= True)

## References

Melbourne Metropolitan Hospitals & Health Services Locations
https://www.healthcollect.vic.gov.au/HospitalLists/MetroHospitalList.aspx

Rural Hospitals & Health Services Locations
https://www.healthcollect.vic.gov.au/HospitalLists/RuralHospitalList.aspx
    
Hospitals Locations in Victoria
https://discover.data.vic.gov.au/dataset/hospital-locations-spatial
